src/frontend/src/components/common/Header/Header.tsx

```diff
@@ -24,27 +25,36 @@ function Header({ user }: Props) {
     localStorage.getItem('access_token'),
   );
   const [showModal, setShowModal] = useState(false);
-  const [userInfo, setUserInfo] = useState({ name: '' });
   const [message, setMessage] = useState('');
   const code = searchParams.get('code');
-
-  const { keycloakToken, error: authCodeError } = useFetchAuthCode(code);
+  const didAuthRef = useRef(false);
   const { keycloakUserDetail, error: userInfoError } =
     useFetchKeycloakUserInfo(accessToken);
 
   // https://github.com/microsoft/TypeScript/issues/48949
   // workaround
   const win: Window = window;
   useEffect(() => {
-    if (keycloakToken !== undefined && keycloakToken !== null) {
-      storeTokens(keycloakToken);
-      setAuthenticated(true);
-
-      setAccessToken(keycloakToken.access_token);
-      navigate(routes.PPQ_LANDING_PAGE);
-    } else return;
+    async function startFetching() {
+      if (didAuthRef.current === false) {
+        try {
+          didAuthRef.current = true;
+          const res = await getAccessToken(code);
+          const keycloakToken = await res?.json();
+          if (keycloakToken !== undefined) {
+            storeTokens(keycloakToken);
+            setAuthenticated(true);
+            setAccessToken(keycloakToken.access_token);
+            navigate(routes.PPQ_LANDING_PAGE);
+          }
+        } catch (e) {
```

Comment:
Critically evaluating -> but I don't feel `ref` are the correct way of implementing it. 
Acc to react documentation as well - `Refs provide a way to access DOM nodes or React elements created in the render method`

And I don't think it makes sense to do another workaround especially when you're reactoring. Happy to hear your side of the argument as well (and links that were followed)

Reply:
that's the trick provided by React team, but let's discuss it later
https://github.com/reactwg/react-18/discussions/18

Reply:
Thank you for the link. Looking into it

Reply:
Now after reading it through. A couple of changes:
One  - Please reevaluate if you even need to use useEffect at first place? I don't see a reason now. All the more cz you no longer have any dependencies.
Two - Even if you do, I would advice using the updated site [beta.react website link]. There is a link at the top of the page you shared. [Here is where it redirects to](https://beta.reactjs.org/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)

Let me know if that makes sense